StartingExample/ExampleMap.cs

```diff
@@ -0,0 +1,34 @@
+﻿using GoRogue;
+using Microsoft.Xna.Framework;
+using SadConsole;
+using System;
+
+namespace StartingExample
+{
+    enum MapLayer
+    {
+        TERRAIN,
+        ITEMS,
+        MONSTERS,
+        PLAYER
+    }
+
+    class ExampleMap : BasicMap
+    {
+        // Handles the changing of tile/entity visiblity as appropriate based on Map.FOV.
+        public FOVVisibilityHandler FovVisibilityHandler { get; }
+
+        public new Player ControlledGameObject
+        {
+            get => (Player)base.ControlledGameObject;
```

Comment:
Since this will throw a cannot cast exception if someone changes the `ControlledGameObject` on `BasicMap`, should we just explicitly test for that and throw an exception that indicates this?

Reply:
In other places like `ComponentBase`, the solution to a similar issue with `Parent` is implemented as virtual properties combined with the `ParentTypeCheck` functionality. Especially given that an event for `ControlledGameObject` changing is mentioned to assist with the memory leak issue, there may be room to bring a similar virtual property and/or type validation system to `ControlledGameObject`?

Reply:
Thinking about it, what is the usefulness of this property besides convenient casting? Is it worth having?

Reply:
It's definitely not strictly necessary - but the idea is basically to make the cast more convenient. The theory is the player class will more often than not contain player-specific data, and as such would need to be casted somewhat frequently during use.  Given that casting is involved in the anticipated use case, the property would be beneficial outside of just the extra convenience if it helped ensure the type safety of that cast (which it currently does not).

In BaseComponent, this is done by having the Parent property fire the added/removed events as appropriate. This allows you can just add an appropriate ParentTypeCheck as a handler to the Added event to enforce that. While this obviously can't catch the issue at compile-time, it does make sure that if you make an error it's made known to you when you set the parent instead of magically crashing when you try to access it (like ControlledGameObject can in it's current state).

Reply:
I implemented this as a type-check so we can see what that looks like, since I think wanting to cast that field is relatively common, and this sets up a way to do implement that desire safely.  If this isn't satisfactory, we can always go back to a simple try-catch that invalid cast on the property; feel free to let me know :D